README
= Description
  A library for playing and controlling sounds on MS Windows.

= Prerequisites
  * ffi
  * test-unit 2 (Development only)

= Installation
  gem install win32-sound

= Synopsis
  require 'win32/sound'
  include Win32

  # Play a wav file
  Sound.play("somefile.wav")

  # Play a system sound
  Sound.play("SystemAsterisk",Sound::ALIAS)

  # Get the current volume of the waveform-audio output device.
  p Sound.volume.join(", ") # left channel, right channel

= Acknowledgements
  API ideas derived (or not) from Perl's Win32::Sound module and Python's
  winsound package.

= Known Bugs
  None that I'm aware of. Please report any bugs on the project page
  at https://github.com/djberg96/win32-sound.
   
= Future Plans
  Add ability to retrieve information about WAV files.
  Add MIDI support?

= Developer's Notes
  The MessageBeep() function, which the Python "winsound" module contains,
  is intentionally omitted here. I felt it was redundant, because you can
  achieve the same effect with something like this:

    Sound.play("SystemAsterisk", Sound::ALIAS).
	
= License
  Artistic 2.0

= Copyright
  (C) 2004-2017, Daniel J. Berger, All Rights Reserved

= Warranty
  This package is provided "as is" and without any express or
  implied warranties, including, without limitation, the implied
  warranties of merchantability and fitness for a particular purpose.

= Author(s)
  Daniel Berger
  Park Heesob
  Dominic Muller


CHANGES
== 0.6.2 - 19-Mar-2017
* Added a win32-sound.rb file for convenience.

== 0.6.1 - 19-Mar-2017
* Added a cert file.
* Removed the .gemtest file.
* All gem related rake tasks now assume Rubygems 2.0 or later.

== 0.6.0 - 30-Jul-2014
* Added the play_freq method. This lets you play a frequency for a
  specified duration at a given volume. Big thanks go out to Dominic
  Muller for the patch.
* Reorganization of FFI related code into their own files.
* Updated the gem:create and gem:install rake tasks.

== 0.5.1 - 10-Apr-2013
* Updated HANDLE prototype in underlying FFI code. This affects
  64 bit versions of Ruby.

== 0.5.0 - 28-Jun-2012
* Now uses FFI instead of win32-api. This means it works with JRuby, too.
* Corresponding updates to the tests, Rakefile, gemspec, etc.
* Removed the Sound::Error exception class. If an underlying FFI function
  fails then a SystemCallError (Errno::XXX) is raised instead.

== 0.4.3 - 21-May-2011
* Refactored Rakefile, removing an old gem install task and adding a
  default task and a clean task.
* Altered the gemspec, including a platform change to 'universal-win32', and
  removed a deprecated specification method.

== 0.4.2 - 6-Aug-2009
* Changed the license to Artistic 2.0.
* Updated to use the Windows::Sound module from the windows-pr library.
* Fixed the Sound::ALIAS value. It was wrongly set to SND_ALIAS_ID.
* Several gemspec updates, including the license and description.
* The Sound.get_wave_volume method is now a true alias.
* Renamed the test file to test_win32_sound.rb.
* Renamed the example file to example_win32_sound.rb to avoid any possible
  confusion with actual test files.
* Added the 'example' rake task.

== 0.4.1 - 27-Jul-2007
* Changed SoundError to Sound::Error.
* Added a Rakefile with tasks for installation and testing.
* Removed the install.rb file. Installation is now handled via a rake task.
* Documentation updates.

== 0.4.0 - 26-Feb-2006
* Now pure Ruby only.
* Documentation updates and reorganization.
* Added a gemspec.
* Minor test suite changes.

== 0.3.0 - 29-May-2005
* Changed the Sound.volume= method (which didn't work) to
  Sound.set_wave_volume because it takes up to two values and only affects
  the wave volume, not the master volume.
* Renamed the Sound.volume method to Sound.wave_volume.
* Fixed a bug in the Sound.wave_volume method.
* Created the Sound.get_wave_volume alias for Sound.wave_volume.
* Fixed the Sound.set_wave_volume method.
* Test adjustments and additions to reflect API changes.
* Now Unicode friendly.
* Removed the sound.rd file.  The sound.txt file is now rdoc friendly.

== 0.2.1 - 1-Mar-2005
* Moved the 'examples' directory to the toplevel directory
* Made the CHANGES and README files rdoc friendly
* Some cleanup in sound.h.

== 0.2.0 - 13-Jul-2004
* Moved the SoundError class under the Win32 module namespace
* Replaced the deprecated STR2CSTR() function with the StringValuePtr()
  function.  This means that as of version 0.2.0, this package requires
  Ruby 1.8.0 or later.
* Moved the test.rb script to doc/examples
* Added the README file (oops).

== 0.1.0 - 14-Feb-2004
* Initial release


test/test_win32_sound.rb
##########################################################################
# test_win32_sound.rb
#
# Test suite for the win32-sound library. You should run this test case
# via the 'rake test' task.
##########################################################################
require 'test-unit'
require 'stringio'
require 'win32/sound'
include Win32

class TC_Win32_Sound < Test::Unit::TestCase
  def setup
    @wav = "c:\\windows\\media\\chimes.wav"
  end

  test "version constant is set to expected value" do
    assert_equal('0.6.2', Sound::VERSION)
  end

  def test_beep
    assert_respond_to(Sound, :beep)
    assert_nothing_raised{ Sound.beep(55, 100) }
  end

  def test_beep_expected_errors
    assert_raises(ArgumentError){ Sound.beep(0, 100) }
    assert_raises(ArgumentError){ Sound.beep }
    assert_raises(ArgumentError){ Sound.beep(500) }
    assert_raises(ArgumentError){ Sound.beep(500, 500, 5) }
  end

  def test_devices
    assert_respond_to(Sound, :devices)
    assert_nothing_raised{ Sound.devices }
    assert_kind_of(Array,Sound.devices)
  end

  def test_stop
    assert_respond_to(Sound, :stop)
    assert_nothing_raised{ Sound.stop }
    assert_nothing_raised{ Sound.stop(true) }
  end

  def test_get_volume_basic
    assert_respond_to(Sound, :wave_volume)
    assert_respond_to(Sound, :get_wave_volume)
    assert_nothing_raised{ Sound.get_wave_volume }
  end

  def test_get_volume
    assert_kind_of(Array, Sound.get_wave_volume)
    assert_equal(2, Sound.get_wave_volume.length)
  end

  def test_set_volume
    assert_respond_to(Sound, :set_wave_volume)
    assert_nothing_raised{ Sound.set_wave_volume(30000) } # About half
    assert_nothing_raised{ Sound.set_wave_volume(30000, 30000) }
  end

  def test_play
    assert_respond_to(Sound, :play)
    assert_nothing_raised{ Sound.play(@wav) }
  end

  def test_play_alias
    assert_nothing_raised{ Sound.play('SystemAsterisk', Sound::ALIAS) }
  end

  def test_expected_errors
    assert_raises(ArgumentError){ Sound.beep(-1, 1) }
  end

  test "play_freq basic functionality" do
    assert_respond_to(Sound, :play_freq)
    assert_nothing_raised{ Sound.play_freq(660.0, 500) }
  end

  test "play_freq raises an error if the frequency is too high or too low" do
    assert_raise(ArgumentError){ Sound.play_freq(999999999, 500) }
    assert_raise(ArgumentError){ Sound.play_freq(1, 500) }
  end

  test "play_freq raises an error if the duration is too high or too low" do
    assert_raise(ArgumentError){ Sound.play_freq(660.0, 9999999) }
    assert_raise(ArgumentError){ Sound.play_freq(660.0, -1) }
  end

  test "play_freq displays a warning if volume is too large" do
    @orig_stderr = $stderr
    $stderr = StringIO.new
    Sound.play_freq(440, 200, 2)
    $stderr.rewind
    assert_equal("WARNING: Volume greater than 1 will cause audio clipping.",
      $stderr.string.chomp)
    $stderr = @orig_stderr
  end

  test "expected constants are defined" do
    assert_not_nil(Sound::ALIAS)
    assert_not_nil(Sound::APPLICATION)
    assert_not_nil(Sound::ASYNC)
    assert_not_nil(Sound::FILENAME)
    assert_not_nil(Sound::LOOP)
    assert_not_nil(Sound::MEMORY)
    assert_not_nil(Sound::NODEFAULT)
    assert_not_nil(Sound::NOSTOP)
    assert_not_nil(Sound::NOWAIT)
    assert_not_nil(Sound::PURGE)
    assert_not_nil(Sound::SYNC)
    assert_not_nil(Sound::VERSION)
    assert_not_nil(Sound::WAVE_FORMAT_PCM)
    assert_not_nil(Sound::WAVE_MAPPER)
    assert_not_nil(Sound::HWAVEOUT)
    assert_not_nil(Sound::WAVEFORMATEX)
    assert_not_nil(Sound::WAVEHDR)
  end

  test "ffi functions are private" do
    assert_not_respond_to(Sound, :Beep)
    assert_not_respond_to(Sound, :PlaySound)
    assert_not_respond_to(Sound, :waveOutSetVolume)
    assert_not_respond_to(Sound, :waveOutGetVolume)
    assert_not_respond_to(Sound, :waveOutGetNumDevs)
    assert_not_respond_to(Sound, :waveInGetNumDevs)
    assert_not_respond_to(Sound, :midiInGetNumDevs)
    assert_not_respond_to(Sound, :midiOutGetNumDevs)
    assert_not_respond_to(Sound, :auxGetNumDevs)
    assert_not_respond_to(Sound, :mixerGetNumDevs)
    assert_not_respond_to(Sound, :waveOutOpen)
    assert_not_respond_to(Sound, :waveOutPrepareHeader)
    assert_not_respond_to(Sound, :waveOutWrite)
    assert_not_respond_to(Sound, :waveOutUnprepareHeader)
    assert_not_respond_to(Sound, :waveOutClose)
  end

  def teardown
    @wav = nil
  end
end


lib/win32/windows/structs.rb
require 'ffi'

module Windows
  module SoundStructs
    extend FFI::Library
    
    # Define an HWAVEOUT struct for use by all the waveOut functions.
    # It is a handle to a waveOut stream, so starting up multiple
    # streams using different handles allows for simultaneous playback.
    # You never need to actually look at the struct, C takes care of
    # its value.
    #
    class HWAVEOUT < FFI::Struct
      layout :i, :int
    end

    # Define WAVEFORMATEX which defines the format (PCM in this case)
    # and various properties like sampling rate, number of channels, etc.
    #
    class WAVEFORMATEX < FFI::Struct
  
      # Initializes struct with sensible defaults for most commonly used
      # values.  While setting these manually is possible, please be
      # sure you know what changes will result in, as an incorrectly
      # set struct will result in unpredictable behavior.
      #
      def initialize(nSamplesPerSec = 44100, wBitsPerSample = 16, nChannels = 1, cbSize = 0)
        self[:wFormatTag] = Win32::Sound::WAVE_FORMAT_PCM
        self[:nChannels] = nChannels
        self[:nSamplesPerSec] = nSamplesPerSec
        self[:wBitsPerSample] = wBitsPerSample
        self[:cbSize] = cbSize
        self[:nBlockAlign] = (self[:wBitsPerSample] >> 3) * self[:nChannels]
        self[:nAvgBytesPerSec] = self[:nBlockAlign] * self[:nSamplesPerSec]
      end
    
      layout(
        :wFormatTag,      :ushort,
        :nChannels,       :ushort,
        :nSamplesPerSec,  :ulong,
        :nAvgBytesPerSec, :ulong,
        :nBlockAlign,     :ushort,
        :wBitsPerSample,  :ushort,
        :cbSize,          :ushort
      )
    end

    #define WAVEHDR which is a header to a block of audio
    #lpData is a pointer to the block of native memory that,
    # in this case, is an integer array of PCM data

    class WAVEHDR < FFI::Struct
    
      # Initializes struct with sensible defaults for most commonly used
      # values.  While setting these manually is possible, please be
      # sure you know what changes will result in, as an incorrectly
      # set struct will result in unpredictable behavior.
      #
      def initialize(lpData, dwBufferLength, dwFlags = 0, dwLoops = 1)
        self[:lpData] = lpData
        self[:dwBufferLength] = dwBufferLength
        self[:dwFlags] = dwFlags
        self[:dwLoops] = dwLoops
      end
      
      layout(
        :lpData,          :pointer,
        :dwBufferLength,  :ulong,
        :dwBytesRecorded, :ulong,
        :dwUser,          :ulong,
        :dwFlags,         :ulong,
        :dwLoops,         :ulong,
        :lpNext,          :pointer,
        :reserved,        :ulong
      )
    end
  end # SoundStructs
end # Windows


lib/win32/windows/functions.rb
require 'ffi'

module Windows
  module SoundFunctions
    extend FFI::Library

    typedef :ulong, :dword
    typedef :uintptr_t, :hmodule
    typedef :uintptr_t, :hwaveout
    typedef :uint, :mmresult

    ffi_lib :kernel32

    private

    # Make FFI functions private
    def self.attach_pfunc(*args)
      attach_function(*args)
      private args[0]
    end

    attach_pfunc :Beep, [:dword, :dword], :bool

    ffi_lib :winmm

    attach_pfunc :PlaySound, [:string, :hmodule, :dword], :bool
    attach_pfunc :waveOutSetVolume, [:hwaveout, :dword], :int
    attach_pfunc :waveOutGetVolume, [:hwaveout, :pointer], :int
    attach_pfunc :waveOutGetNumDevs, [], :int
    attach_pfunc :waveInGetNumDevs, [], :int
    attach_pfunc :midiOutGetNumDevs, [], :int
    attach_pfunc :midiInGetNumDevs, [], :int
    attach_pfunc :auxGetNumDevs, [], :int
    attach_pfunc :mixerGetNumDevs, [], :int
    attach_pfunc :waveOutOpen, [:pointer, :uint, :pointer, :dword, :dword, :dword], :mmresult
    attach_pfunc :waveOutPrepareHeader, [:hwaveout, :pointer, :uint], :mmresult
    attach_pfunc :waveOutWrite, [:hwaveout, :pointer, :uint], :mmresult
    attach_pfunc :waveOutUnprepareHeader, [:hwaveout, :pointer, :uint], :mmresult
    attach_pfunc :waveOutClose, [:hwaveout], :mmresult
  end
end


examples/example_win32_sound.rb
#######################################################################
# example_win32_sound.rb (win32-sound)
#
# A example script to demonstrate the win32-sound library and for
# general futzing. You can run this via the 'rake example' task.
#
# Modify as you see fit.
#######################################################################
require 'win32/sound'
include Win32

wav = "c:\\windows\\media\\chimes.wav"

puts "VERSION: " + Sound::VERSION
#puts "Devices: " + Sound.devices.join(", ")

#Sound.volume = [77,128] # my personal settings

orig_left, orig_right = Sound.wave_volume
puts "Volume was: #{orig_left}, #{orig_right}"

#Sound.volume = 140
#puts "Volume is now: " + Sound.volume.join(", ")

#Sound.volume = [orig_left,orig_right]
#puts "Volume is now: " + Sound.volume.join(", ")

puts "Playing 'SystemAsterisk' sound"
sleep 1
Sound.play("SystemAsterisk",Sound::ALIAS)

puts "Playing 'chimes' sound once"
sleep 1
Sound.play(wav)

puts "Playing 'chimes' sound in a loop for 3 seconds"
sleep 1
Sound.play(wav,Sound::ASYNC|Sound::LOOP)
sleep 3
Sound.stop

puts "Playing default sound"
sleep 1
Sound.play("Foofoo", Sound::ALIAS)

puts "Playing a beep"
sleep 1
Sound.beep(500, 2000)

puts "Playing multiple frequencies simultaneously"
sleep 1
threads = []
[440, 660].each do |freq|
  threads << Thread.new { Sound.play_freq(freq) }
end
threads.each { |th| th.join }


lib/win32/sound.rb
require_relative 'windows/functions'
require_relative 'windows/structs'

# The Win32 module serves as a namespace only.
module Win32

  # The Sound class encapsulates various methods for playing sound as well
  # as querying or configuring sound related properties.
  class Sound
    include Windows::SoundStructs
    extend Windows::SoundFunctions

    # The version of the win32-sound library
    VERSION = '0.6.2'.freeze

    private

    LOW_FREQUENCY   = 37
    HIGH_FREQUENCY  = 32767
    MAX_VOLUME      = 0xFFFF
    WAVE_FORMAT_PCM = 1     # Waveform-audio data is PCM.
    WAVE_MAPPER     = -1    # Used by waveOutOpen.  The function selects
                            # a waveform-audio output device capable of
                            # playing the given format.

    public

    SYNC            = 0x00000000 # play synchronously (default)
    ASYNC           = 0x00000001 # play asynchronously
    NODEFAULT       = 0x00000002 # silence (!default) if sound not found
    MEMORY          = 0x00000004 # pszSound points to a memory file
    LOOP            = 0x00000008 # loop the sound until next sndPlaySound
    NOSTOP          = 0x00000010 # don't stop any currently playing sound
    NOWAIT          = 8192       # don't wait if the driver is busy
    ALIAS           = 65536      # name is a registry alias
    ALIAS_ID        = 1114112    # alias is a predefined ID
    FILENAME        = 131072     # name is file name
    RESOURCE        = 262148     # name is resource name or atom
    PURGE           = 0x00000040 # purge non-static events for task
    APPLICATION     = 0x00000080 # look for app specific association

    # Plays a frequency for a specified duration at a given volume.
    # Defaults are 440Hz, 1 second, full volume.
    #
    # The result is a single channel, 44100Hz sampled, 16 bit sine wave.
    # If multiple instances are plays in simultaneous threads, they
    # will be started and played at the same time.
    #
    # ex.: threads = []
    #      [440, 660].each do |freq|
    #        threads << Thread.new { Win32::Sound.play_freq(freq) }
    #      end
    #      threads.each { |th| th.join }
    #
    # The first frequency in this array (440) will wait until the
    # thread for 660 finished calculating its PCM array and they
    # will both start streaming at the same time.
    #
    # If immediate_playback is set to false, the thread will calculate
    # all pending PCM arrays and wait to be woken up again.  This
    # if useful for time-sensitive playback of notes in succession.
    #
    def self.play_freq(frequency = 440, duration = 1000, volume = 1, immediate_playback = true)

      if frequency > HIGH_FREQUENCY || frequency < LOW_FREQUENCY
        raise ArgumentError, 'invalid frequency'
      end

      if duration < 0 || duration > 5000
        raise ArgumentError, 'invalid duration'
      end

      if volume.abs > 1
        warn("WARNING: Volume greater than 1 will cause audio clipping.")
      end

      stream(immediate_playback) do |wfx|
        data = generate_pcm_integer_array_for_freq(frequency, duration, volume)
        data_buffer = FFI::MemoryPointer.new(:int, data.size)
        data_buffer.write_array_of_int data
        buffer_length = wfx[:nAvgBytesPerSec]*duration/1000
        WAVEHDR.new(data_buffer, buffer_length)
      end

    end

    # Returns an array of all the available sound devices; their names contain
    # the type of the device and a zero-based ID number. Possible return values
    # are WAVEOUT, WAVEIN, MIDIOUT, MIDIIN, AUX or MIXER.
    #
    def self.devices
      devs = []

      begin
        0.upto(waveOutGetNumDevs()){ |i| devs << "WAVEOUT#{i}" }
        0.upto(waveInGetNumDevs()){ |i| devs << "WAVEIN#{i}" }
        0.upto(midiOutGetNumDevs()){ |i| devs << "MIDIOUT#{i}" }
        0.upto(midiInGetNumDevs()){ |i| devs << "MIDIIN#{i}" }
        0.upto(auxGetNumDevs()){ |i| devs << "AUX#{i}" }
        0.upto(mixerGetNumDevs()){ |i| devs << "MIXER#{i}" }
      rescue Exception
        raise SystemCallError, FFI.errno, "GetNumDevs"
      end

      devs
    end

    # Generates simple tones on the speaker. The function is synchronous; it
    # does not return control to its caller until the sound finishes.
    #
    # The frequency (in Hertz) must be between 37 and 32767.
    # The duration is in milliseconds.
    #
    def self.beep(frequency, duration)
      if frequency > HIGH_FREQUENCY || frequency < LOW_FREQUENCY
        raise ArgumentError, 'invalid frequency'
      end

      if 0 == Beep(frequency, duration)
        raise SystemCallError, FFI.errno, "Beep"
      end

      self
    end

    # Stops any currently playing waveform sound. If +purge+ is set to
    # true, then *all* sounds are stopped. The default is false.
    #
    def self.stop(purge = false)
      if purge && purge != 0
        flags = PURGE
      else
        flags = 0
      end

      unless PlaySound(nil, 0, flags)
        raise SystemCallError, FFI.errno, "PlaySound"
      end

      self
    end

    # Plays the specified sound. The sound can be a wave file or a system
    # sound, when used in conjunction with the ALIAS flag.
    #
    # Valid flags:
    #
    # Sound::ALIAS
    #   The sound parameter is a system-event alias in the registry or the
    #   WIN.INI file. If the registry contains no such name, it plays the
    #   system default sound unless the NODEFAULT value is also specified.
    #   Do not use with FILENAME.
    #
    # Sound::APPLICATION
    #   The sound is played using an application-specific association.
    #
    # Sound::ASYNC
    #   The sound is played asynchronously and the function returns
    #   immediately after beginning the sound.
    #
    # Sound::FILENAME
    #   The sound parameter is the name of a WAV file.  Do not use with
    #   ALIAS.
    #
    # Sound::LOOP
    #   The sound plays repeatedly until Sound.stop() is called. You must
    #   also specify the ASYNC flag to loop sounds.
    #
    # Sound::MEMORY
    #   The sound points to an image of a waveform sound in memory.
    #
    # Sound::NODEFAULT
    #   If the sound cannot be found, the function returns silently without
    #   playing the default sound.
    #
    # Sound::NOSTOP
    #   If a sound is currently playing, the function immediately returns
    #   false without playing the requested sound.
    #
    # Sound::NOWAIT
    #   If the driver is busy, return immediately without playing the sound.
    #
    # Sound::PURGE
    #   Stop playing all instances of the specified sound.
    #
    # Sound::SYNC
    #   The sound is played synchronously and the function does not return
    #   until the sound ends.
    #
    # Examples:
    #
    #    require 'win32/sound'
    #    include Win32
    #
    #    # Play a wave file once
    #    Sound.play('some_file.wav')
    #
    #    # Play a wave file in an asynchronous loop for 2 seconds
    #    Sound.play('some_file.wav', Sound::ASYNC | Sound::LOOP)
    #    sleep 2
    #    Sound.stop
    #
    def self.play(sound, flags = 0)
      unless PlaySound(sound, 0, flags)
        raise SystemCallError, FFI.errno, "PlaySound"
      end

      self
    end

    # Sets the volume for the left and right channel. If the +right_channel+
    # is omitted, the volume is set for *both* channels.
    #
    # You may optionally pass a single Integer rather than an Array, in which
    # case it is assumed you are setting both channels to the same value.
    #
    def self.set_wave_volume(left_channel, right_channel = nil)
      right_channel ||= left_channel

      lvolume = left_channel > MAX_VOLUME ? MAX_VOLUME : left_channel
      rvolume = right_channel > MAX_VOLUME ? MAX_VOLUME : right_channel

      volume = lvolume | rvolume << 16

      if waveOutSetVolume(-1, volume) != 0
        raise SystemCallError, FFI.errno, "waveOutSetVolume"
      end

      self
    end

    # Returns a 2-element array that contains the volume for the left channel
    # and right channel, respectively.
    #
    def self.wave_volume
      ptr = FFI::MemoryPointer.new(:ulong)

      if waveOutGetVolume(-1, ptr) != 0
        raise SystemCallError, FFI.errno, "waveOutGetVolume"
      end

      volume = ptr.read_long

      [low_word(volume), high_word(volume)]
    end

    class << self
      alias get_wave_volume wave_volume
    end

    private

    def self.low_word(num)
      num & 0xFFFF
    end

    def self.high_word(num)
      num >> 16
    end

    # Sets up a ready-made waveOut stream to push a PCM integer array to.
    # It expects a block to be associated with the method call to which
    # it will yield an instance of WAVEFORMATEX that the block uses
    # to prepare a WAVEHDR to return to the function.
    #
    # The WAVEHDR can contain either a self-made PCM integer array
    # or an array from a wav file or some other audio file converted
    # to PCM.
    #
    # This function will take the entire PCM array and create one
    # giant buffer, so it is not intended for audio streams larger
    # than 5 seconds.
    #
    # In order to play larger audio files, you will have to use the waveOut
    # functions and structs to set up a double buffer to incrementally
    # push PCM data to.
    #
    def self.stream(immediate_playback)
      hWaveOut = HWAVEOUT.new
      wfx = WAVEFORMATEX.new

      wfx[:wFormatTag] = WAVE_FORMAT_PCM
      wfx[:nChannels] = 1
      wfx[:nSamplesPerSec] = 44100
      wfx[:wBitsPerSample] = 16
      wfx[:cbSize] = 0
      wfx[:nBlockAlign] = (wfx[:wBitsPerSample] >> 3) * wfx[:nChannels]
      wfx[:nAvgBytesPerSec] = wfx[:nBlockAlign] * wfx[:nSamplesPerSec]

      if ((error_code = waveOutOpen(hWaveOut.pointer, WAVE_MAPPER, wfx.pointer, 0, 0, 0)) != 0)
        raise SystemCallError.new('waveOutOpen', FFI.errno)
      end

      header = yield(wfx)

      if ((error_code = waveOutPrepareHeader(hWaveOut[:i], header.pointer, header.size)) != 0)
        raise SystemCallError.new('waveOutPrepareHeader', FFI.errno)
      end

      unless immediate_playback
        Thread.stop
        Thread.current[:sleep_time] ||= 0
        sleep Thread.current[:sleep_time]
      end
      Thread.pass

      if (waveOutWrite(hWaveOut[:i], header.pointer, header.size) != 0)
        raise SystemCallError.new('waveOutWrite', FFI.errno)
      end

      while (waveOutUnprepareHeader(hWaveOut[:i], header.pointer, header.size) == 33)
        sleep 0.1
      end

      if ((error_code = waveOutClose(hWaveOut[:i])) != 0)
        raise SystemCallError.new('waveOutClose', FFI.errno)
      end

      self
    end

    # Generates an array of PCM integers to play a particular frequency
    # It also ramps up and down the volume in the first and last
    # 200 milliseconds to prevent audio clicking.
    #
    def self.generate_pcm_integer_array_for_freq(freq, duration, volume)
      data = []
      ramp = 200.0
      samples = (44100/2*duration/1000.0).floor

      samples.times do |sample|

        angle = (2.0*Math::PI*freq) * sample/samples * duration/1000
        factor = Math.sin(angle)
        x = 32768.0*factor*volume

        if sample < ramp
          x *= sample/ramp
        end
        if samples - sample < ramp
          x *= (samples - sample)/ramp
        end

        data << x.floor
      end

      data
    end
  end # Sound
end # Win32
